File: problems/tier1-foundation/008-file-search/solution.go
// File Search — Strategy + Composite reference solution (Go).
package main

import "sort"

type FileNode struct {
	name        string
	size        int
	extension   string
	isDirectory bool
	children    []*FileNode
}

type SearchCriteria interface {
	matches(f *FileNode) bool
}

type SearchByExtension struct{ ext string }

func (s SearchByExtension) matches(f *FileNode) bool {
	return !f.isDirectory && f.extension == s.ext
}

type SearchByMinSize struct{ minSize int }

func (s SearchByMinSize) matches(f *FileNode) bool {
	return !f.isDirectory && f.size >= s.minSize
}

type SearchByName struct{ substring string }

func (s SearchByName) matches(f *FileNode) bool {
	return contains(f.name, s.substring)
}

func contains(s, sub string) bool {
	if sub == "" {
		return true
	}
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

type AndFilter struct{ criteria []SearchCriteria }

func (a AndFilter) matches(f *FileNode) bool {
	for _, c := range a.criteria {
		if !c.matches(f) {
			return false
		}
	}
	return true
}

type OrFilter struct{ criteria []SearchCriteria }

func (o OrFilter) matches(f *FileNode) bool {
	for _, c := range o.criteria {
		if c.matches(f) {
			return true
		}
	}
	return false
}

func dfs(node *FileNode, criteria SearchCriteria, results *[]*FileNode) {
	if node == nil {
		return
	}
	if !node.isDirectory && criteria.matches(node) {
		*results = append(*results, node)
	}
	for _, c := range node.children {
		dfs(c, criteria, results)
	}
}

func searchByExtension(root *FileNode, ext string) []*FileNode {
	out := []*FileNode{}
	dfs(root, SearchByExtension{ext}, &out)
	return out
}

func searchBySize(root *FileNode, minSize int) []*FileNode {
	out := []*FileNode{}
	dfs(root, SearchByMinSize{minSize}, &out)
	return out
}

func searchByName(root *FileNode, substring string) []*FileNode {
	out := []*FileNode{}
	dfs(root, SearchByName{substring}, &out)
	return out
}

func searchComposite(root *FileNode, criteria []SearchCriteria, mode string) []*FileNode {
	var f SearchCriteria
	if mode == "AND" {
		f = AndFilter{criteria}
	} else {
		f = OrFilter{criteria}
	}
	out := []*FileNode{}
	dfs(root, f, &out)
	return out
}

func searchAndSort(root *FileNode, criteria SearchCriteria, sortBy string) []*FileNode {
	out := []*FileNode{}
	dfs(root, criteria, &out)
	switch sortBy {
	case "name":
		sort.SliceStable(out, func(i, j int) bool { return out[i].name < out[j].name })
	case "size":
		sort.SliceStable(out, func(i, j int) bool { return out[i].size > out[j].size })
	case "extension":
		sort.SliceStable(out, func(i, j int) bool { return out[i].extension < out[j].extension })
	}
	return out
}

// ─── Module fixture state ───────────────────────────────────────────────────

var gRoot *FileNode

func reset_service() {
	gRoot = nil
}

func fs_build_default_tree() {
	mainCpp := &FileNode{name: "main.cpp", size: 50, extension: "cpp"}
	utilsCpp := &FileNode{name: "utils.cpp", size: 120, extension: "cpp"}
	helperH := &FileNode{name: "helper.h", size: 10, extension: "h"}
	readme := &FileNode{name: "readme.md", size: 5, extension: "md"}
	report := &FileNode{name: "report.pdf", size: 200, extension: "pdf"}
	buildSh := &FileNode{name: "build.sh", size: 2, extension: "sh"}
	src := &FileNode{name: "src", isDirectory: true, children: []*FileNode{mainCpp, utilsCpp, helperH}}
	docs := &FileNode{name: "docs", isDirectory: true, children: []*FileNode{readme, report}}
	gRoot = &FileNode{name: "project", isDirectory: true, children: []*FileNode{src, docs, buildSh}}
}

func fs_build_empty_tree() {
	gRoot = &FileNode{name: "empty", isDirectory: true}
}

func fs_build_single_file_tree() {
	f := &FileNode{name: "test.txt", size: 30, extension: "txt"}
	gRoot = &FileNode{name: "root", isDirectory: true, children: []*FileNode{f}}
}

func fs_count_by_extension(ext string) int {
	if gRoot == nil {
		return 0
	}
	return len(searchByExtension(gRoot, ext))
}

func fs_has_by_extension(ext string, name string) bool {
	if gRoot == nil {
		return false
	}
	for _, f := range searchByExtension(gRoot, ext) {
		if f.name == name {
			return true
		}
	}
	return false
}

func fs_count_by_size(minSize int) int {
	if gRoot == nil {
		return 0
	}
	return len(searchBySize(gRoot, minSize))
}

func fs_has_by_size(minSize int, name string) bool {
	if gRoot == nil {
		return false
	}
	for _, f := range searchBySize(gRoot, minSize) {
		if f.name == name {
			return true
		}
	}
	return false
}

func fs_count_by_name(sub string) int {
	if gRoot == nil {
		return 0
	}
	return len(searchByName(gRoot, sub))
}

func fs_has_by_name(sub string, name string) bool {
	if gRoot == nil {
		return false
	}
	for _, f := range searchByName(gRoot, sub) {
		if f.name == name {
			return true
		}
	}
	return false
}

func fs_count_composite_and(ext string, minSize int) int {
	if gRoot == nil {
		return 0
	}
	return len(searchComposite(gRoot, []SearchCriteria{SearchByExtension{ext}, SearchByMinSize{minSize}}, "AND"))
}

func fs_count_composite_or(ext string, minSize int) int {
	if gRoot == nil {
		return 0
	}
	return len(searchComposite(gRoot, []SearchCriteria{SearchByExtension{ext}, SearchByMinSize{minSize}}, "OR"))
}

func fs_first_sorted_by(ext string, sortBy string) string {
	if gRoot == nil {
		return ""
	}
	v := searchAndSort(gRoot, SearchByExtension{ext}, sortBy)
	if len(v) > 0 {
		return v[0].name
	}
	return ""
}

File: problems/tier1-foundation/008-file-search/solution.py
"""File Search — Strategy + Composite + Decorator reference solution (Python)."""

from abc import ABC, abstractmethod


class FileNode:
    def __init__(self, name, size, extension, isDirectory, children=None):
        self.name = name
        self.size = size
        self.extension = extension
        self.isDirectory = isDirectory
        self.children = children if children else []


class SearchCriteria(ABC):
    @abstractmethod
    def matches(self, file):
        ...


class SearchByExtension(SearchCriteria):
    def __init__(self, ext):
        self.ext = ext

    def matches(self, file):
        return (not file.isDirectory) and file.extension == self.ext


class SearchByMinSize(SearchCriteria):
    def __init__(self, minSize):
        self.minSize = minSize

    def matches(self, file):
        return (not file.isDirectory) and file.size >= self.minSize


class SearchByName(SearchCriteria):
    def __init__(self, substring):
        self.substring = substring

    def matches(self, file):
        return self.substring in file.name


class AndFilter(SearchCriteria):
    def __init__(self, criteria):
        self.criteria = criteria

    def matches(self, file):
        return all(c.matches(file) for c in self.criteria)


class OrFilter(SearchCriteria):
    def __init__(self, criteria):
        self.criteria = criteria

    def matches(self, file):
        return any(c.matches(file) for c in self.criteria)


def _dfs(node, criteria, results):
    if node is None:
        return
    if not node.isDirectory and criteria.matches(node):
        results.append(node)
    for c in node.children:
        _dfs(c, criteria, results)


def search_by_extension(root, ext):
    out = []
    _dfs(root, SearchByExtension(ext), out)
    return out


def search_by_size(root, minSize):
    out = []
    _dfs(root, SearchByMinSize(minSize), out)
    return out


def search_by_name(root, substring):
    out = []
    _dfs(root, SearchByName(substring), out)
    return out


def search_composite(root, criteria, mode):
    f = AndFilter(criteria) if mode == "AND" else OrFilter(criteria)
    out = []
    _dfs(root, f, out)
    return out


def search_and_sort(root, criteria, sortBy):
    out = []
    _dfs(root, criteria, out)
    if sortBy == "name":
        out.sort(key=lambda f: f.name)
    elif sortBy == "size":
        out.sort(key=lambda f: -f.size)
    elif sortBy == "extension":
        out.sort(key=lambda f: f.extension)
    return out


# ─── Module fixture state ───────────────────────────────────────────────────

_g_root = None


def reset_service():
    global _g_root
    _g_root = None


def fs_build_default_tree():
    global _g_root
    main_cpp = FileNode("main.cpp", 50, "cpp", False)
    utils_cpp = FileNode("utils.cpp", 120, "cpp", False)
    helper_h = FileNode("helper.h", 10, "h", False)
    readme = FileNode("readme.md", 5, "md", False)
    report = FileNode("report.pdf", 200, "pdf", False)
    build_sh = FileNode("build.sh", 2, "sh", False)
    src = FileNode("src", 0, "", True, [main_cpp, utils_cpp, helper_h])
    docs = FileNode("docs", 0, "", True, [readme, report])
    _g_root = FileNode("project", 0, "", True, [src, docs, build_sh])


def fs_build_empty_tree():
    global _g_root
    _g_root = FileNode("empty", 0, "", True)


def fs_build_single_file_tree():
    global _g_root
    f = FileNode("test.txt", 30, "txt", False)
    _g_root = FileNode("root", 0, "", True, [f])


def fs_count_by_extension(ext):
    return len(search_by_extension(_g_root, ext)) if _g_root else 0


def fs_has_by_extension(ext, name):
    if _g_root is None:
        return False
    return any(f.name == name for f in search_by_extension(_g_root, ext))


def fs_count_by_size(minSize):
    return len(search_by_size(_g_root, minSize)) if _g_root else 0


def fs_has_by_size(minSize, name):
    if _g_root is None:
        return False
    return any(f.name == name for f in search_by_size(_g_root, minSize))


def fs_count_by_name(sub):
    return len(search_by_name(_g_root, sub)) if _g_root else 0


def fs_has_by_name(sub, name):
    if _g_root is None:
        return False
    return any(f.name == name for f in search_by_name(_g_root, sub))


def fs_count_composite_and(ext, minSize):
    if _g_root is None:
        return 0
    return len(search_composite(_g_root, [SearchByExtension(ext), SearchByMinSize(minSize)], "AND"))


def fs_count_composite_or(ext, minSize):
    if _g_root is None:
        return 0
    return len(search_composite(_g_root, [SearchByExtension(ext), SearchByMinSize(minSize)], "OR"))


def fs_first_sorted_by(ext, sortBy):
    if _g_root is None:
        return ""
    v = search_and_sort(_g_root, SearchByExtension(ext), sortBy)
    return v[0].name if v else ""
